internal/oci/mutate/map.go

```diff
@@ -0,0 +1,167 @@
+// that returns IsAfterChildren(ctx) == true.
+// If the returned entity is nil, it is filtered from the result of Map.
+type Mutator func(context.Context, oci.SignedEntity) (oci.SignedEntity, error)
+
+// ErrSkipChildren is a special error that may be returned from a Mutator
+// to skip processing of an index's child entities.
+var ErrSkipChildren = errors.New("skip child entities")
+
+// Map calls `fn` on the signed entity and each of its constituent entities (`SignedImageIndex`
+// or `SignedImage`) transitively.
+// Any errors returned by an `fn` are returned by `Map`.
+func Map(ctx context.Context, parent oci.SignedEntity, fn Mutator) (oci.SignedEntity, error) {
+	parent, err := fn(before(ctx), parent)
+	switch {
+	case errors.Is(err, ErrSkipChildren):
+		return parent, nil
+	case err != nil:
+		return nil, err
+	case parent == nil:
+		// If the function returns nil, it filters it.
+		return nil, nil
+	}
+
+	sii, ok := parent.(oci.SignedImageIndex)
+	if !ok {
+		return parent, nil
+	}
+	im, err := sii.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	// Track whether any of the child entities change.
+	changed := false
+
+	adds := []IndexAddendum{}
+	for _, desc := range im.Manifests {
+		switch desc.MediaType {
+		case types.OCIImageIndex, types.DockerManifestList:
+			x, err := sii.SignedImageIndex(desc.Digest)
+			if err != nil {
+				return nil, err
+			}
+
+			se, err := Map(ctx, x, fn)
+			if err != nil {
+				return nil, err
+			} else if se == nil {
+				// If the function returns nil, it filters it.
+				changed = true
+				continue
+			}
+
+			changed = changed || (x != se)
+			adds = append(adds, IndexAddendum{
+				Add: se.(oci.SignedImageIndex), // Must be an image index.
+				Descriptor: v1.Descriptor{
+					URLs:        desc.URLs,
+					MediaType:   desc.MediaType,
+					Annotations: desc.Annotations,
+					Platform:    desc.Platform,
+				},
+			})
+
+		case types.OCIManifestSchema1, types.DockerManifestSchema2:
+			x, err := sii.SignedImage(desc.Digest)
+			if err != nil {
+				return nil, err
+			}
+
+			se, err := fn(ctx, x)
+			if err != nil {
+				return nil, err
+			} else if se == nil {
+				// If the function returns nil, it filters it.
+				changed = true
+				continue
+			}
+
+			changed = changed || (x != se)
+			adds = append(adds, IndexAddendum{
+				Add: se.(oci.SignedImage), // Must be an image
+				Descriptor: v1.Descriptor{
+					URLs:        desc.URLs,
+					MediaType:   desc.MediaType,
+					Annotations: desc.Annotations,
+					Platform:    desc.Platform,
+				},
+			})
+
+		default:
+			return nil, fmt.Errorf("unknown mime type: %v", desc.MediaType)
+		}
+	}
+
+	if !changed {
+		return parent, nil
+	}
+
+	// Construct a new ImageIndex from the new consituent signed images.
+	result := AppendManifests(mutate.IndexMediaType(empty.Index, im.MediaType), adds...)
+
+	// Since the children changed, give the callback a crack at the new image index.
+	return fn(after(ctx), result)
```

Comment:
Should make sure we're also preserving manifest-level annotations: https://github.com/opencontainers/image-spec/blob/main/annotations.md

Reply:
IDK if we have a function for this in GGCR, lemme look.

At a minimum I'll TODO it.

Reply:
Ok, I found `mutate.Annotations` that has a weird interface, but it gets the job done.